### ISSUE TITLE
i10n_gen errors/warnings improvement suggestions + need testing

### ISSUE THREAD

**Issue by zanderso** [flutter/flutter#115501]
~~After the change here: https://github.com/flutter/flutter/pull/112390, building the new Flutter gallery for iOS fails at tip-of-tree~~. [eta: the build failure was a different issue, filed a separate issue.]

I see some issues with https://github.com/flutter/flutter/pull/112390.
1. Lots of commented out code. If a fast-follow had been landed as @HansMuller mentioned, then that would have been okay, but it's been 11 days or so.
2. I'm not seeing a test for this particular error path: `The plural part...`.
3. The tests that do exist, don't seem to be testing that the second two lines of the error message print something useful.
4. The second part of the message is coming through empty for me, and the third part, which is supposed to have the file and line number are empty, so I can't track down the issue with flutter/gallery.
5. nit: As mentioned on the PR, "overrided" should be "overridden".

cc @thkim1011 @HansMuller @christopherfujino 

There's a separate issue, which seems to be that the new Xcode error handling code seems to treat `printWarning`s from i10n_gen as fatal, both with and without https://github.com/flutter/flutter/pull/112390. cc @jmagman

**Comment by thkim1011** [comment]
For the first point, I'm almost done with the follow up PR, should have a PR out end of week. Regarding the logger warning, I can just remove it as Xcode seems to be handling warnings as errors; does that sound good?

**Comment by christopherfujino** [comment]
As a side point, I wonder why this is not failing on our https://ci.chromium.org/p/flutter/builders/prod/Mac_ios%20new_gallery_ios__transition_perf or https://ci.chromium.org/p/flutter/builders/prod/Mac_ios%20new_gallery_impeller_ios__transition_perf devicelab tests?

Or by "tip of tree", do you mean the tip of the new flutter gallery repo, which hasn't rolled to flutter tests yet?

**Comment by jmagman** [comment]
> There's a separate issue, which seems to be that the new Xcode error handling code seems to treat `printWarning`s from i10n_gen as fatal,

I am on ToT 26a990f610ce460e2bd067358185142a65990639 and ToT https://github.com/flutter/gallery/commit/add9d71d2a8193b6d0ec01ec6e8b09c74759d547.

I see the i10n_gen warnings, but only the first time I build after a `flutter clean`:
```
"vi": 6 untranslated message(s).
"zh": 6 untranslated message(s).
"zh_CN": 6 untranslated message(s).
"zh_HK": 6 untranslated message(s).
"zh_TW": 6 untranslated message(s).
"zu": 6 untranslated message(s).
To see a detailed report, use the untranslated-messages-file
option in the l10n.yaml file:
untranslated-messages-file: desiredFileName.txt
<other option>: <other selection>


This will generate a JSON format file containing all messages that
need to be translated.
...
Xcode build done.                                            3.0s
Failed to build iOS app
Error (Xcode): "Runner" requires a provisioning profile. Select a provisioning profile in the Signing & Capabilities editor.
/Users/m/Projects/gallery/ios/Runner.xcodeproj
```

`"Runner" requires a provisioning profile` error is expected because the gallery is using manual codesigning, and I don't have the provisioning profile set up on my machine.  When I open the Xcode project and update the codesigning in "Signing & Capabilities" to work on my machine, then `flutter clean` again to get the warnings,  it builds successfully.

@zanderso could you file a new issue with `-v` output to track just the Xcode `printWarning` part of your issue, since I'm not able to reproduce?

**Comment by zanderso** [comment]
@jmagman I filed https://github.com/flutter/flutter/issues/115567. More detailed repro steps there, but if you use a simulator instead of hardware, then the build won't get blocked on the provisioning profile check.

@thkim1011 I think there is a general usability issue around the messages issued by the i10n_gen code. Firstly there are way too many in flutter/gallery. That is a problem if we think that flutter/gallery is a representative usage of our i10n code. Second, the warnings don't give enough information to fix the problem, and don't appear to say where or what the problem is. Whether they're errors or warnings is kind of a side issue.

**Comment by thkim1011** [comment]
Whoops, I actually fixed most of the exceptions to show exactly which file and message the error occurs in but I think I missed this because it was a warning. I can add this in a quick PR.

As to why flutter/gallery is throwing these warnings has something to do with how the internal translation console's plural syntax is subtly different from ICU's syntax which is implemented in flutter. TC assumes that the =1 case is distinct from the "one" case (I don't exactly understand why that is), but they are not different in ICU. When the .arb file gets generated from TC, it includes both the =1 and "one" case, but I think this warrants a warning because this implies that one of these cases is just dead code.

I can also fix flutter/gallery's internal tool for generating the .arb files so that it doesn't include this redundancy.

**Comment by jmagman** [comment]
> More detailed repro steps there, but if you use a simulator instead of hardware, then the build won't get blocked on the provisioning profile check.

I didn't get blocked, I never saw the error you saw, the build succeeded.  I'll follow up in https://github.com/flutter/flutter/issues/115567

**Comment by jonahwilliams** [comment]
I'm also not able to build https://github.com/gskinnerTeam/flutter-wonderous-app/ anymore

```
[   +1 ms] Generating synthetic localizations package failed with 1 error:
           
           Exception: Make sure that the specified placeholder is defined in your arb file.
           Explore details about {title}.
```

I'm not familiar with the arb format/syntax, so these could be errors ...

**Comment by jonahwilliams** [comment]
also, no idea what file this is caused by

**Comment by jonahwilliams** [comment]
also ... I can only get the errors one at a time.

**Comment by thkim1011** [comment]
https://github.com/gskinnerTeam/flutter-wonderous-app/blob/master/lib/l10n/app_en.arb

@jonahwilliams every message needs a corresponding JSON object with metadata about the message describing what the placeholders are. I'm a bit surprised that the old parser and code generator did not check for this.
See [here](https://docs.google.com/document/d/10e0saTfAv32OZLRmONy866vnaw0I2jwL8zukykpgWBc/edit).

**Comment by zanderso** [comment]
@thkim1011 In order to allow users to make a smooth transition when upgrading Flutter, is it possible for the tool to assume some reasonable defaults? Maybe that was the old behavior? Either way, this change in behavior seems like a breaking change, and so it will need a migration guide before it can go into a release.

**Comment by HansMuller** [comment]
I agree. Although this change does correct an old parsing error, the fact the developers have relied on it means that it's effectively a breaking change. Maybe we should provide a way to enable the old behavior, even if that's not the default, and provide a short migration guide and a helpful error message that steer developers in the right direction.

**Comment by thkim1011** [comment]
Just ran the Flutter wonderous app on 3.3.9. The generated files look like this:
```
import 'app_localizations.dart';

/// The translations for English (`en`).
class AppLocalizationsEn extends AppLocalizations {
  AppLocalizationsEn([String locale = 'en']) : super(locale);

  @override
  String get appName => 'Wonderous';

  @override
  String get localeSwapButton => '简体中文';

  @override
  String get animatedArrowSemanticSwipe => 'Explore details about {title}.';

  @override
  String get appBarTitleFactsHistory => 'Facts and History';

  @override
  String get appBarTitleConstruction => 'Construction';

  @override
  String get appBarTitleLocation => 'Location Info';

  @override
  String get bottomScrubberSemanticScrubber => 'scrubber';

  @override
  String get bottomScrubberSemanticTimeline => 'Timeline Scrubber, drag horizontally to navigate the timeline.';

  @override
  String get collectionLabelDiscovered => '{percentage}% discovered';

  @override
  String get collectionLabelCount => '{count} of {total}';
```
It seems as if placeholders are completely ignored if they're not explicitly specified in the `app_en.arb` template file. It shouldn't be too hard to add this behavior as an option for backward compatibility. Does this sound like a good solution?

**Comment by jonahwilliams** [comment]
Thank you for double checking that - I think that is a reasonable approach to warn for now, something like:

```
Warning: Missing placeholder declaration in foo.arb L43:

"message": "Explore details about {title}",
                                                           ^

To declare a place holder, the following thingy would need to be added to foo.arb:

"@message": {
  "description":  "...",
  "placeholders": {
    "title": {
      "type": “String”,
      “example”: “...”
    }
  }
}

In a future release of Flutter, this will become a compilation error.

```

i.e. should include the file and line where the problem was, how to fix it, and that it will become an error in the future.

**Comment by esDotDev** [comment]
Curious the rationale for forcing developers to do this? 

I10n already gets deprioritized by teams because it is too much work. Forcing every single key to have a description, that is often 3-4x as long as the key itself, makes this worse not better. 

I understand the need to enforce certain standards, but this feels like making perfect the enemy of good. We've been localizing apps for 20 years using google spreadsheets, descriptions were always possible, occasionally they made sense to use, often they did not. 

Why not leave the decision in the hands of the people with the most information regarding the various stakeholders and available budget?

**Comment by esDotDev** [comment]
Also, fwiw, we have odd error, is `plural` a reserved keyword or something?
```
"newlyDiscoveredLabelNew": "{count} new item{plural} to explore",
"@newlyDiscoveredLabelNew": {"placeholders": {"count": {}, "plural": {}}},
```
== 
`ICU Syntax Error: Expected "identifier" but found "plural"`

If I change the param name to something else it will compile.

**Comment by esDotDev** [comment]
Another issue that has cropped up with the latest changes: we have certain cases where we need to replace a param with an entire widget, not a string.

For example:
`"aboutUsText": "Built with {flutterUrl} by the team at {companyUrl}."`

In this case, due to how Flutter handles text links, we need to split this into 4 text-spans so we can re-combine them for the appearance of hyperlinks:
![image](https://user-images.githubusercontent.com/736973/204695482-e15ffbfd-5d62-40ad-b72e-d1541c51a647.png)

In HTML we could just replace with an `href` but I'm not quite sure how to make this work with this new system and it's extra forced structure.

**Comment by thkim1011** [comment]
Hey @esDotDev, thanks for the suggestions. I like the idea of making the placeholder/description metadata optional. I've added it into my PR. Also, I fixed the issue with "plural" being a keyword; initially it was like this to easily tokenize the messages, but I found a way to get around it.



> Another issue that has cropped up with the latest changes: we have certain cases where we need to replace a param with an entire widget, not a string.
> 
> For example: `"aboutUsText": "Built with {flutterUrl} by the team at {companyUrl}."`
> 
> In this case, due to how Flutter handles text links, we need to split this into 4 text-spans so we can re-combine them for the appearance of hyperlinks: ![image](https://user-images.githubusercontent.com/736973/204695482-e15ffbfd-5d62-40ad-b72e-d1541c51a647.png)
> 
> In HTML we could just replace with an `href` string but I'm not quite sure how to make this work with this new system and it's extra forced structure. I supposed I could inject the same keys back in, and then parse the text manually as we were doing before, but seems pretty smelly:
> 
> ```
> final String aboutUs = strings.aboutUsText('flutterUrl', 'companyUrl');
> // build spans...
> ```
> 
> Not a big deal, bought thought I would surface in case it sparks some ideas.
> 
> This is a case where having optional param declarations (and corresponding method signature) would be beneficial as we wouldn't need to make this extra step.
Is this a new issue with the latest changes? I'm not too sure how the l10n changes might have changed the process for adding hyperlinks within the `Text` widget.

**Comment by esDotDev** [comment]
They only changed it, in that it now assumes all params are going to be replaced, but in cases like this, we actually want to use them to `split` the string, so we can perform more work on it. So the assumption of replacement forces an extra step.

```dart
splitOnSomeToken(strings.aboutUs, { 'flutterUrl': 'http://...'});

vs

splitOnSomeToken(strings.aboutUs('flutterUrl'), { 'flutterUrl': 'http://...'});
```
This extra `strings.aboutUs('flutterUrl')` is forced, when we just want to do `strings.aboutUs` and work on the string from there. Here we just re-inject the same key, for lack of a better solution, which works but not ideal. Would be nice if we can leave off param declerations, which signals that we want to use a `get` method.

**Comment by thkim1011** [comment]
Are you saying that you would like to use special characters `{` and `}` in a message string without parsing it as a parameter? I added an escaping syntax a few days ago, see #116137.

**Comment by esDotDev** [comment]
Ya, I'm basically asking for any `{param}` that is not declared, would be ignored.

Though now that you mention it we could just use a different type of brackets for tokens that we want to split on, vs ones we want to replace, for example we could use `[]` instead, and then it would be ignored automatically.
